File: c++/p001-p100/leetcode024.cpp
// <!-- encoding UTF-8 --!>
/*****************************************************************************
*                     ----Stay Hungry Stay Foolish----                       *
*   @author :   Shen                                                         *
*   @name   :   Leetcode 024 Swap Nodes in Pairs                             *
*****************************************************************************/

#include <bits/stdc++.h>
using namespace std;
typedef long long int64;
template<class T>inline bool updateMin(T& a, T b){ return a > b ? a = b, 1: 0; }
template<class T>inline bool updateMax(T& a, T b){ return a < b ? a = b, 1: 0; }

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *swapPairs(ListNode *head)
    {
        ListNode **hoge = &head;
        ListNode *piyo;
        ListNode *fuga;
        while ((piyo = *hoge) && (fuga = piyo->next))
        {
            piyo->next = fuga->next;
            fuga->next = piyo;
            *hoge = fuga;
            hoge = &(piyo->next);
        }
        return head;
    }
};

void test() {}

int main() { test(); return 0; }

File: Archive#00-001-100/024-Swap Nodes in Pairs/leetcode024.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val)
#     @val = val
#     @next = nil
#   end
# end

# @param {ListNode} head
# @return {ListNode}
def swap_pairs(head)
  dummy = ListNode.new(0)
  dummy.next = head
  fuga, hoge = dummy, head
  # before: fuga -> [hoge -> piyo] -> ...
  # after:  fuga -> [piyo -> hoge] -> ...
  while (hoge != nil) do
    if hoge.next == nil then
      break
    end
    piyo = hoge.next
    fuga.next = piyo
    hoge.next = piyo.next
    piyo.next = hoge
    fuga = hoge
    hoge = hoge.next
  end
  dummy.next
end